rpc/rpc.go

```diff
@@ -1256,245 +1249,234 @@ func (c *Conn) handleRelease(ctx context.Context, id exportID, count uint32) err
-		}
-		content, err := ans.results.Content()
-		if err != nil {
-			c.mu.Unlock()
-			return failedf("incoming disembargo: read answer ID %d: %v", tgt.promisedAnswer, err)
-		}
-		ptr, err := capnp.Transform(content, tgt.transform)
+		var (
+			imp    *importClient
+			client *capnp.Client
+		)
+
+		syncutil.With(&c.mu, func() {
+			if tgt.which != rpccp.MessageTarget_Which_promisedAnswer {
+				err = rpcerr.Failedf("incoming disembargo: sender loopback: target is not a promised answer")
+				return
+			}
+
+			ans := c.answers[tgt.promisedAnswer]
+			if ans == nil {
+				err = rpcerr.Failedf("incoming disembargo: unknown answer ID %d", tgt.promisedAnswer)
+				return
+			}
+			if ans.flags&returnSent == 0 {
+				err = rpcerr.Failedf("incoming disembargo: answer ID %d has not sent return", tgt.promisedAnswer)
+				return
+			}
+
+			if ans.err != nil {
+				err = rpcerr.Failedf("incoming disembargo: answer ID %d returned exception", tgt.promisedAnswer)
+				return
+			}
+
+			var content capnp.Ptr
+			if content, err = ans.results.Content(); err != nil {
+				err = rpcerr.Failedf("incoming disembargo: read answer ID %d: %v", tgt.promisedAnswer, err)
+				return
+			}
+
+			ptr, err := capnp.Transform(content, tgt.transform)
+			if err != nil {
+				err = rpcerr.Failedf("incoming disembargo: read answer ID %d: %v", tgt.promisedAnswer, err)
+				return
+			}
+
+			iface := ptr.Interface()
+			if !iface.IsValid() || int64(iface.Capability()) >= int64(len(ans.resultCapTable)) {
+				err = rpcerr.Failedf("incoming disembargo: sender loopback requested on a capability that is not an import")
+				return
+			}
+
+			client := ans.resultCapTable[iface.Capability()].AddRef()
+
+			var ok bool
+			syncutil.Without(&c.mu, func() {
+				imp, ok = client.State().Brand.Value.(*importClient)
+			})
+
+			if !ok || imp.c != c {
+				client.Release()
+				err = rpcerr.Failedf("incoming disembargo: sender loopback requested on a capability that is not an import")
+				return
+			}
+
+			// TODO(maybe): check generation?
+		})
+
 		if err != nil {
-			c.mu.Unlock()
-			return failedf("incoming disembargo: read answer ID %d: %v", tgt.promisedAnswer, err)
-		}
-		iface := ptr.Interface()
-		if !iface.IsValid() || int64(iface.Capability()) >= int64(len(ans.resultCapTable)) {
-			c.mu.Unlock()
-			return failedf("incoming disembargo: sender loopback requested on a capability that is not an import")
-		}
-		client := ans.resultCapTable[iface.Capability()].AddRef()
-		c.mu.Unlock()
-		imp, ok := client.State().Brand.Value.(*importClient)
-		c.mu.Lock()
-		if !ok || imp.c != c {
-			c.mu.Unlock()
-			client.Release()
-			return failedf("incoming disembargo: sender loopback requested on a capability that is not an import")
+			release()
+			return err
 		}
-		// TODO(maybe): check generation?
 
 		// Since this Cap'n Proto RPC implementation does not send imports
 		// unless they are fully dequeued, we can just immediately loop back.
```

Comment:
You didn't add this but: I have no idea what this comment is trying to say, do you? (It sounds like it documents some important invariant, and I am therefore uncomfortable with the fact that I don't understand it).

Reply:
Best guess:  this has to do with some kind of "version" of something?  Grepping through the C++ implementation reveals use of the term "generation" consistent with this meaning (see [here](https://github.com/capnproto/capnproto/search?q=generation)).  Notably, it seems to involve reconnection of some sort, so I suppose the "generation" refers to the number of prior connections, or something of the sort.

Does that ring any bells?